### ISSUE TITLE
An explanation of "how to train policy (agent)" such as `Basic_DQN` would be valuable

### ISSUE THREAD

**Issue by JinraeKim** [JuliaReinforcementLearning/ReinforcementLearning.jl#240]
I'm trying to implement my RL algorithm but it is hard to understand how to do it.
It may be a silly question. Sorry for frequently asking how to use this repo.

For example, I took a look at [`basic_dqn.jl`](https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl/blob/master/src/algorithms/dqns/basic_dqn.jl).

From the following code, I guess that there is stored data (trajectory), sample data from batch using `BatchSampler`, and take gradient update.
```julia
function RLBase.update!(learner::BasicDQNLearner, traj::AbstractTrajectory)
    if length(traj) >= learner.min_replay_history
        inds, batch = sample(learner.rng, traj, learner.sampler)
        update!(learner, batch)
    end
end

function RLBase.update!(learner::BasicDQNLearner, batch::NamedTuple{SARTS})

    Q = learner.approximator
    γ = learner.γ
    loss_func = learner.loss_func

    s, a, r, t, s′ = send_to_device(device(Q), batch)
    a = CartesianIndex.(a, 1:length(a))

    gs = gradient(params(Q)) do
        q = Q(s)[a]
        q′ = vec(maximum(Q(s′); dims = 1))
        G = r .+ γ .* (1 .- t) .* q′
        loss = loss_func(G, q)
        ignore() do
            learner.loss = loss
        end
        loss
    end

    update!(Q, gs)
end
```

However, I cannot figure out
1. where is the `BatchSampler` defined?
2. when and where is the `RLBase.update!` called? (this function is for training approximator, right?)
It would be important for some offline algorithms, such as [behaviour cloning (BC) for discrete action](https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl/blob/master/src/algorithms/offline_rl/behavior_cloning.jl). Note that I would implement BC algorithm for deterministic MDP with continuous state and action spaces (this would merely be an easy application of supervised learning) after understanding the full workflow of this repo :)
3. If so, how can I evaluate the trained agent as described [here](https://juliareinforcementlearning.org/guide/#how_to_evaluate_an_agent_during_training)?
4. `send_to_device` seems related to GPU support then how should I change the "device"?
5. What is `t` in `s, a, r, t, s'`?  ~~If it means the (discrete-time) step,~~ It seems that `t` denotes a binary of termination according to [this](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/c14ec876311408096a677f1b998e307cafff4dc4/src/policies/agents/trajectories/abstract_trajectory.jl). I would like to recommend you to change this symbol.
It is because `t` usually denotes "continuous time" and discretised time step (namely, `n`) corresponds to `n t_diff = t` where `t_diff` denotes the length of time step. Also, `t` is already used for step or count in hooks.
Of course, it's merely a recommendation :)

+) 
As a long term plan, I carefully recommend you to provide a introductory RL theory and example code.
It is because RL is a too broad research field so there are many confusion in terminology, problem setting, etc.
I think that [Spinning up RL](https://spinningup.openai.com/en/latest/) is a good example.

**Comment by findmyway** [comment]
First, thanks for your valuable feedback and suggestions.

1. BatchSampler is defined in RLCore https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/c14ec876311408096a677f1b998e307cafff4dc4/src/policies/agents/trajectories/trajectory_extension.jl

2. Update may happen at many different stages based on the implementation detail of the specific algorithm. by default it is called in PreActStage . I added some description here https://juliareinforcementlearning.org/blog/an_introduction_to_reinforcement_learning_jl_design_implementations_thoughts/#21_the_general_workflow

3. If you’ve finished reading the above section , you’ll see that evaluating agent is simply `run(agent.policy, env, stop_condition, hook)`

4. `send_to_device` is meant to be used internal only to support different DNN framework or devices. Unfortunately we only support Flux with CUDA now. So just use the `gpu ` function in Flux to move data/model between cpu and gpu

5. I totally agree. `t` is reserved for historical reasons. I can’t find a general variable name for the termination flag. Peope use different variable names in many popular packages. Please recommend one if you have a better suggestion.

The blog above was meant to unify the symbols used across different RL subfields and this package. However, I'm under a lot of stress at work recently and can not find enough time to finish that blog. I hope oneday we have enough active contributors and we can work on it togethr.

**Comment by JinraeKim** [comment]
> First, thanks for your valuable feedback and suggestions.
> 
> 1. BatchSampler is defined in RLCore https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/c14ec876311408096a677f1b998e307cafff4dc4/src/policies/agents/trajectories/trajectory_extension.jl
> 2. Update may happen at many different stages based on the implementation detail of the specific algorithm. by default it is called in PreActStage . I added some description here https://juliareinforcementlearning.org/blog/an_introduction_to_reinforcement_learning_jl_design_implementations_thoughts/#21_the_general_workflow
> 3. If you’ve finished reading the above section , you’ll see that evaluating agent is simply `run(agent.policy, env, stop_condition, hook)`
> 4. `send_to_device` is meant to be used internal only to support different DNN framework or devices. Unfortunately we only support Flux with CUDA now. So just use the `gpu ` function in Flux to move data/model between cpu and gpu
> 5. I totally agree. `t` is reserved for historical reasons. I can’t find a general variable name for the termination flag. Peope use different variable names in many popular packages. Please recommend one if you have a better suggestion.
> 
> The blog above was meant to unify the symbols used across different RL subfields and this package. However, I'm under a lot of stress at work recently and can not find enough time to finish that blog. I hope oneday we have enough active contributors and we can work on it togethr.

Thank you a lot.

(To 1-4) Ok, I'll take a look at it. Thank you!
(To 5.) Well... I prefer using `is_terminated` but it's too long and I'm not sure whether it's widely accepted. Also, it is already used as a function. Sorry.

**Comment by JinraeKim** [comment]
@findmyway 
Sorry for annoying questions :(

(Related to the questions 2, 3)
According to the [blog](https://juliareinforcementlearning.org/blog/an_introduction_to_reinforcement_learning_jl_design_implementations_thoughts/#21_the_general_workflow), there are some `update` methods corresponding to stages such as
![image](https://user-images.githubusercontent.com/43136096/115685200-fb346600-a392-11eb-9795-c586d18e1ad5.png)

However, the [Behaviour Cloning code](https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl/blob/master/src/algorithms/offline_rl/behavior_cloning.jl) defines `RLBase.update!(p::BehaviorCloningPolicy, batch::NamedTuple{(:state, :action)})`.
I'm little bit confused since `RLBase.update!(p, batch)` does not match with the blog.

Also, suppose that somehow the policy is updated at some stages.
Then, how can I avoid the updates when using hooks for evaluation?
If a policy is evaluated through a hook, then it seems that the policy would be updated at the stages I've supposed.

**Comment by findmyway** [comment]
> However, the Behaviour Cloning code defines RLBase.update!(p::BehaviorCloningPolicy, batch::NamedTuple{(:state, :action)}).
I'm little bit confused since RLBase.update!(p, batch) does not match with the blog.

It might not be that straightforward  here.

Those update functions you mentioned in the blog are very general ones. When designing/implementing new algorithms we usually narrow down the context to lower level objects. For example the named tuple of batch experience here. (And it is encouraged to do so to make the underlying components more reusable)

In the current design, if one wants to update policy during interactions with environment,it is recommended to wrap the policy in `Agent`. In evaluation mode the policy is detached from the `Agent` and is assumed to not be changed. However, since everything is transparent, you can still do whatever you want with it.

**Comment by JinraeKim** [comment]
> > However, the Behaviour Cloning code defines RLBase.update!(p::BehaviorCloningPolicy, batch::NamedTuple{(:state, :action)}).
> > I'm little bit confused since RLBase.update!(p, batch) does not match with the blog.
> 
> It might not be that straightforward here.
> 
> Those update functions you mentioned in the blog are very general ones. When designing/implementing new algorithms we usually narrow down the context to lower level objects. For example the named tuple of batch experience here. (And it is encouraged to do so to make the underlying components more reusable)
> 
> In the current design, if one wants to update policy during interactions with environment,it is recommended to wrap the policy in `Agent`. In evaluation mode the policy is detached from the `Agent` and is assumed to not be changed. However, since everything is transparent, you can still do whatever you want with it.

@findmyway Please let me know if there's something that I understand wrongly. Thank you.

Ok...
Relying on your explanation, I've investigated codes and finally get it probably.
Note that [a new blog link](https://juliareinforcementlearning.org/blog/an_introduction_to_reinforcement_learning_jl_design_implementations_thoughts/#21_the_general_workflow) is very good for understanding this package.

## My understanding about `Agent` and policy
So `Agent` is a wrapper for policy. When training a policy, I may be able to use a wrapper `Agent`.
In evaluation mode, I can evaluate a (trained) policy by detaching the policy as e.g. `Agent.policy` and simply inject this object to `run` with a hook. Probably it looks like `run(agent.policy, env, stop_condition_eval, hook_eval)` (in the evaluation hook) whereas the entire training code looks like `run(agent, env, stop_condition, hook)`.
Note that the source codes about `Agent` are probably [this](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/master/src/policies/agents/agent.jl).

## My guess about the way of training policy
For example, an RL algorithm (actually, imitation learning) "Behaviour cloning (BC)" is provided [in RLZoo](https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl/blob/master/src/algorithms/offline_rl/behavior_cloning.jl#L24). Here, we can see that there is a method with form `RLBase.update!(policy, trajectory)`.
From [agent.jl in RLCore](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/c14ec876311408096a677f1b998e307cafff4dc4/src/policies/agents/agent.jl#L37), the agent will do something within the loop of an experiment.
By default, "trajectory update" do nothing and "policy update" dispatches `update!(policy, trajectory)` at `PreActStage` according to [docstring](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/c14ec876311408096a677f1b998e307cafff4dc4/src/policies/agents/agent.jl#L64).
Honestly, I don't fully understand where and how the dispatch works.
Anyway, it corresponds to the [batch update of BC algorithm](https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl/blob/022c1fd433911dcaedf3ff38b4bfb6351c544497/src/algorithms/offline_rl/behavior_cloning.jl#L24), which receives arguments of BC poilcy and batch trajectory.

## Remaining questions
1. Where does `update!` for `Agent` at `PreActStage` dispatched for `update!(policy, trajectory)`?
2. If [BC algorithm](https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl/blob/master/src/algorithms/offline_rl/behavior_cloning.jl#L24) is an offline algorithm, the policy might be trained without interaction with environments. In this case, how should I update it? Call the `update!` function with a while loop directly? (Well, it may work...!)

## Issue report
+) There is a typo [here](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/c14ec876311408096a677f1b998e307cafff4dc4/src/policies/agents/agent.jl#L65): it should be "dispatched".

**Comment by findmyway** [comment]
> ## My understanding about `Agent` and policy
> So `Agent` is a wrapper for policy. When training a policy, I may be able to use a wrapper `Agent`.
> In evaluation mode, I can evaluate a (trained) policy by detaching the policy as e.g. `Agent.policy` and simply inject this object to `run` with a hook. Probably it looks like `run(agent.policy, env, stop_condition_eval, hook_eval)` (in the evaluation hook) whereas the entire training code looks like `run(agent, env, stop_condition, hook)`.
> Note that the source codes about `Agent` are probably [this](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/master/src/policies/agents/agent.jl).

You are absolutely right here!

> ## My guess about the way of training policy
> For example, an RL algorithm (actually, imitation learning) "Behaviour cloning (BC)" is provided [in RLZoo](https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl/blob/master/src/algorithms/offline_rl/behavior_cloning.jl#L24). Here, we can see that there is a method with form `RLBase.update!(policy, trajectory)`.
> From [agent.jl in RLCore](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/c14ec876311408096a677f1b998e307cafff4dc4/src/policies/agents/agent.jl#L37), the agent will do something within the loop of an experiment.
> By default, "trajectory update" do nothing and "policy update" dispatches `update!(policy, trajectory)` at `PreActStage` according to [docstring](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/c14ec876311408096a677f1b998e307cafff4dc4/src/policies/agents/agent.jl#L64).
> Honestly, I don't fully understand where and how the dispatch works.
> Anyway, it corresponds to the [batch update of BC algorithm](https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl/blob/022c1fd433911dcaedf3ff38b4bfb6351c544497/src/algorithms/offline_rl/behavior_cloning.jl#L24), which receives arguments of BC poilcy and batch trajectory.
> 
> ## Remaining questions
> 1. Where is `update!` for `Agent` at `PreActStage` dispatched for `update!(policy, trajectory)`?

It's [here](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/c14ec876311408096a677f1b998e307cafff4dc4/src/policies/agents/agent.jl#L74).

And for each specific algorithm, a more specific method is implemented:

https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl/search?p=2&q=PreActStage

> 2. If [BC algorithm](https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl/blob/master/src/algorithms/offline_rl/behavior_cloning.jl#L24) is an offline algorithm, the policy might be trained without interaction with environments. In this case, how should I update it? Call the `update!` function with a while loop directly? (Well, it may work...!) Or, just running an experiment with unnecessary interaction with environments for lazy people.

Yep. Just call `update!` within a while loop directly. And inject any logging functions in it if you like. There's really no need to running an experiment again.

> ## Issue report
> +) There is a typo [here](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/c14ec876311408096a677f1b998e307cafff4dc4/src/policies/agents/agent.jl#L65): it should be "dispatched".

**Comment by JinraeKim** [comment]
> > If [BC algorithm](https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl/blob/master/src/algorithms/offline_rl/behavior_cloning.jl#L24) is an offline algorithm, the policy might be trained without interaction with environments. In this case, how should I update it? Call the `update!` function with a while loop directly? (Well, it may work...!) Or, just running an experiment with unnecessary interaction with environments for lazy people.
> 
> Yep. Just call `update!` within a while loop directly. And inject any logging functions in it if you like. There's really no need to running an experiment again.

Still I don't get it.
I mean, as your example codes of the dispatched method for specific algorithms (e.g., [DDPG](https://github.com/JuliaReinforcementLearning/ReinforcementLearningZoo.jl/blob/bc64e422fe720b5040fbeb8f27cc62f45b1e894a/src/algorithms/policy_gradient/ddpg.jl#L126)), it receives arguments of `RLBase.update!(policy, traj, env, stage)`.
However, the [docstring](https://github.com/JuliaReinforcementLearning/ReinforcementLearningCore.jl/blob/c14ec876311408096a677f1b998e307cafff4dc4/src/policies/agents/agent.jl#L65) says to me that it receives like `RLBase.update!(policy, trajectory)`.
Is it something deprecated? Or, did I miss something?

**Comment by findmyway** [comment]
> However, the docstring says to me that it receives like RLBase.update!(policy, trajectory).
Is it something deprecated? Or, did I miss something?

Ah, yes, it is deprecated in the latest release.

**Comment by findmyway** [comment]
Before v0.8, we mainly use `RLBase.update!(policy, trajectory)`.

**Comment by JinraeKim** [comment]
> > However, the docstring says to me that it receives like RLBase.update!(policy, trajectory).
> > Is it something deprecated? Or, did I miss something?
> 
> Ah, yes, it is deprecated in the latest release.

Everything becomes so peaceful :) Really thank you for your kind replies.

**Comment by findmyway** [comment]
My pleasure!
